src/core/server/saved_objects/migrationsv2/model.ts

```diff
@@ -0,0 +1,730 @@
+        // step).
+        throwBadResponse(stateP, left as never);
+      } else {
+        throwBadResponse(stateP, left);
+      }
+    }
+  } else if (stateP.controlState === 'MARK_VERSION_INDEX_READY_CONFLICT') {
+    // If another instance completed a migration from the same source we need
+    // to check that the completed migration was performed by a Kibana that's
+    // on the same version as this instance.
+    const res = resW as ExcludeRetryableEsError<ResponseType<typeof stateP.controlState>>;
+    if (Either.isRight(res)) {
+      const indices = res.right;
+      const aliases = getAliases(indices);
+      if (
+        aliases[stateP.currentAlias] != null &&
+        aliases[stateP.versionAlias] != null &&
+        aliases[stateP.currentAlias] === aliases[stateP.versionAlias]
+      ) {
+        // If the current and version aliases are pointing to the same index
+        // the migration was completed by another instance on the same version
+        // and it's safe to start serving traffic.
+        return { ...stateP, controlState: 'DONE' };
+      } else {
+        // Fail the migration, the instance that completed the migration is
+        // running a different version of Kibana. This avoids a situation where
+        // we loose acknowledged writes because two versions are both
+        // accepting writes, but are writing into difference indices.
+        const conflictingKibanaVersion =
+          indexVersion(aliases[stateP.currentAlias]) ?? aliases[stateP.currentAlias];
+        return {
+          ...stateP,
+          controlState: 'FATAL',
+          reason: `Multiple versions of Kibana are attempting a migration in parallel. Another Kibana instance on version ${conflictingKibanaVersion} completed this migration (this instance is running ${stateP.kibanaVersion}). Ensure that all Kibana instances are running on same version and try again.`,
+        };
+      }
+    } else {
+      throwBadResponse(stateP, res);
+    }
+  } else if (stateP.controlState === 'DONE' || stateP.controlState === 'FATAL') {
+    // The state-action machine will never call the model in the terminating states
+    throwBadControlState(stateP as never);
+  } else {
+    return throwBadControlState(stateP);
+  }
+};
+
+/**
+ * Construct the initial state for the model
+ */
+export const createInitialState = ({
+  kibanaVersion,
+  targetMappings,
+  preMigrationScript,
+  migrationVersionPerType,
+  indexPrefix,
+}: {
+  kibanaVersion: string;
+  targetMappings: IndexMapping;
+  preMigrationScript?: string;
+  migrationVersionPerType: SavedObjectsMigrationVersion;
+  indexPrefix: string;
+}): InitState => {
+  const outdatedDocumentsQuery = {
+    bool: {
+      should: Object.entries(migrationVersionPerType).map(([type, latestVersion]) => ({
+        bool: {
+          must: { term: { type } },
+          must_not: { term: { [`migrationVersion.${type}`]: latestVersion } },
+        },
+      })),
+    },
+  };
```

Comment:
yeah @restrry had a good idea to create a `KibanaModel` that could contain all the state and logic that's static, things we don't collect as part of the algorithm but rather that's injected into it at the start. We could still attach the KibanaModel onto the state, but it could for instance be a class and wouldn't have to be a POJO like the rest of the state.

Didn't get around to it though :(